cli/popper/runner_host.py

```diff
@@ -36,225 +38,215 @@ def run(self, step):
+                                             self._config.workspace_dir,
+                                             self._spawned_pids)
+        if pid != 0:
+            self._spawned_pids.remove(pid)
+
         return ecode
 
     def stop_running_tasks(self):
-        for p in HostRunner.spawned_processes:
-            log.info(f'Stopping proces {p.pid}')
-            p.kill()
+        for pid in self._spawned_pids:
+            log.info(f'Stopping proces {pid}')
+            os.kill(pid, signal.SIGKILL)
+
+    @staticmethod
+    def _exec_cmd(cmd, env=None, cwd=os.getcwd(), pids=set(), logging=True):
+        pid = 0
+        try:
+            with Popen(cmd, stdout=PIPE, stderr=STDOUT,
+                       universal_newlines=True, preexec_fn=os.setsid,
+                       env=env, cwd=cwd, text=True) as p:
+                pid = p.pid
+                pids.add(p.pid)
+                log.debug('Reading process output')
+
+                output = []
+                for line in iter(p.stdout.readline, ''):
+                    if logging:
+                        log.step_info(line)
+                    else:
+                        output.append(line)
+
+                p.wait()
+                ecode = p.poll()
+
+            log.debug(f'Code returned by process: {ecode}')
+
+        except SubprocessError as ex:
+            output = ""
+            ecode = ex.returncode
+            log.step_info(f"Command '{cmd[0]}' failed with: {ex}")
+        except Exception as ex:
+            output = ""
+            ecode = 1
+            log.step_info(f"Command raised non-SubprocessError error: {ex}")
+
+        return pid, ecode, '\n'.join(output)
 
 
 class DockerRunner(StepRunner):
-    """Runs steps in docker."""
-    d = None
+    """Runs steps in docker on the local machine."""
+    def __init__(self, init_docker_client=True, **kw):
+        super(DockerRunner, self).__init__(**kw)
 
-    # hold references to spawned containers
-    spawned_containers = []
+        self._spawned_containers = []
+        self._d = None
 
-    def __init__(self, config):
-        super(DockerRunner, self).__init__(config)
+        if not init_docker_client:
+            return
 
         try:
-            DockerRunner.d = docker.from_env()
-            DockerRunner.d.version()
+            self._d = docker.from_env()
+            self._d.version()
         except Exception as e:
             log.debug(f'Docker error: {e}')
             log.fail(f'Unable to connect to the docker daemon.')
 
-        log.debug(f'Docker info: {pu.prettystr(DockerRunner.d.info())}')
+        log.debug(f'Docker info: {pu.prettystr(self._d.info())}')
 
     def __exit__(self, exc_type, exc_value, exc_traceback):
-        if DockerRunner.d:
-            DockerRunner.d.close()
-        DockerRunner.spawned_containers = []
+        if self._d:
+            self._d.close()
+        self._spawned_containers = []
         return True
```

Comment:
We don't need this `return=True` statement right ?

Reply:
true, I'll remove

Reply:
can you add this on your end? in that way I avoid disrupting the state of this branch on your end

Reply:
yes. sure.